### PR TITLE
Tag and check for release candidates

### LINKED ISSUE #60: automation: dealing with rejections
Sometimes packages are rejected for reasons that are not picked up by rhub or CI runs (example: URL changes from http -> https).

In this case, `check_release_state()` switches back from "submitted" to "ready-to-release". However, no immediate release is desired because some fixes are necessary.

How do we detect this special case?

### PR DESCRIPTION
**Pull request by pat-s**
fixes #60 

- Creates RC tags in `release()`
- Aborts the build if  the build is running in "state = release" and the sha of the most recent commit == sha of the most recent tag && "-rc" is found in the most recent tag name
- Deletes all RC tags in `post_release()`
- Uses `gh` (if avail) to merge a PR instead of merging the branch.

## ToDo

- [x] Enumerate RC tags because we can have multiple rejections

### CODE REVIEW

**Comment by pat-s** [comment]
@krlmlr Successfully used this branch to release mlr v2.18.0 including to failed attempt (which triggered two "-rc" tags).

**Comment by krlmlr** [comment]
I'm moving towards bumping the version with each attempted release. It's also recommended practice by CRAN.

**Comment by pat-s** [comment]
Not a friend of this practice.

Does this mean you do not want to merge this PR at all?

**Comment by krlmlr** [comment]
Numbers are cheap.

We might support both ways, it's low priority at the moment.

**Comment by pat-s** [comment]
I don't mind the numbers per se (I'd there is a need for a new release) but the inconsistency in news numbering (personal splean probably).

**Comment by pat-s** [comment]
Here's a snapshot how `unleash("patch")` looks like currently (using the `tag_release_candidates_gert` branch.

<details>
<summary>pre_release()</summary>

image

</details>

<details>
<summary>release()</summary>

image


</details>

<details>
<summary>post_release()</summary>

image


</details>

File: R/auto.R [hunk 1]
```diff
@@ -325,13 +334,36 @@ post_release_impl <- function() {
...
-  # FIXME: Check if PR open, if yes merge PR instead
-  release_branch <- get_branch_name()
-  switch_branch(get_main_branch())
-  merge_branch(release_branch)
-  push_head(get_head_branch())
+  # FIXME: integrate `gh` system lib dep
+  # FIXME: clean up mess
+
+  # check if gh is avail
+  is_gh_avail <- suppressWarnings(try(system2("gh", stdout = FALSE, stderr = FALSE), silent = TRUE))
```
Comment (krlmlr) [inline R/auto.R:345]: I'd rather discuss this change in a separate PR.

**Review by krlmlr — CHANGES_REQUESTED** [review]
Thanks. Can you please isolate the changes so that they apply directly to the `f-27-automate` branch, and open a new PR? I'm not ready to commit to state management, RC tagging is useful in itself.

File: R/auto.R (hunk 1 above)
Comment (krlmlr) [reply R/auto.R:345]: Separate = third.

File: R/auto.R [hunk 2]
```diff
@@ -64,10 +64,12 @@ pre_release_impl <- function(which, force) {
   # push main branch, bump to devel version and push again
   push_to_new(remote_name, force)
   switch_branch(main_branch)
+  push_head(main_branch)
```
Comment (pat-s) [inline R/auto.R:67]: not sure if these two `push_head()` calls are still needed - guess we need to check in action

**Comment by pat-s** [comment]
Rebased and stripped the diff, not sure if I removed something important though. 

I'd just check how it behaves in practice then but ofc this is your decision in the end.

**Comment by pat-s** [comment]
Did your latest comment have the intention to merge this PR?

**Comment by krlmlr** [comment]
No.

**Comment by krlmlr** [comment]
I'm experimenting with a flow that allows seamless integration of the target branch into the source branch. I do wonder why GitHub hasn't implemented this yet. Anyway, this is now available through the "/ merge" command in pull requests. Because GHA won't run checks when I push directly, I open a PR instead so that the user can merge manually and trigger checks in the target branch.

CC @wael-sadek.

**Comment by pat-s** [comment]
@krlmlr How do you want to proceed here? I assume closing?

**Comment by krlmlr** [comment]
I'm open to discussing release candidate tags. About to merge that long-running PR, this will probably close this PR.

**Pull request closed without merging**